sky/clouds/gcp.py

```diff
@@ -524,6 +524,92 @@ def get_vcpus_mem_from_instance_type(
         return service_catalog.get_vcpus_mem_from_instance_type(instance_type,
                                                                 clouds='gcp')
 
+    def get_available_reservation_resources(
+        self,
+        instance_type: str,
+        region: str,
+        zone: Optional[str],
+        specific_reservations: Set[str],
+    ) -> int:
+        del region  # Unused
+        assert zone is not None, 'GCP requires zone to get available reservations.'
+        reservations = self._list_reservations_for_instance_type_in_zone(
+            instance_type, zone)
+
+        return self._count_available_resources_for_reservations(
+            reservations, specific_reservations)
+
+    def _list_reservations_for_instance_type_in_zone(
+        self,
+        instance_type: str,
+        zone: str,
+    ) -> List[Dict]:
+        reservations = self._list_reservations_for_instance_type(instance_type)
+        return [r for r in reservations if r['zone'].endswith(f'/{zone}')]
+
+    @functools.lru_cache()
+    def _list_reservations_for_instance_type(
+        self,
+        instance_type: str,
+    ) -> List[Dict]:
```

Comment:
I am thinking more about this. This will cause the `--retry-until-up` always use the same list of reservations, which will cause the information to be stale after multiple retries. We may need a timeout for the cache to be cleaned. The following should work:

```suggestion
    @functools.lru_cache(maxsize=1)
    def _list_reservations_for_instance_type(
        self,
        instance_type: str,
        ttl_hash: int,
    ) -> List[Dict]:
        del ttl_hash # Unused
```

When calling this function we can pass the ttl_hash with:
```python
# The time of the reservation cache live. It can be updated based
# on the users' request.
RESERVATION_INFO_TTL = 300 # seconds
self._list_reservations_for_instance_type(instance_type, ttl_hash=int(time.time() / RESERVATION_INFO_TTL))
```

What do you think?

Reply:
I used the TLRUCache from cachetools instead. It avoid polluting the method interface with a ttl hash.